### PR TITLE
School safety zones

### LINKED ISSUE #511: Updating file paths in the school safety zones DAG
Updating the path to `schools.py` script which is imported within vz_google_sheets DAG.
Old path: `/etc/airflow/data_scripts/vision_zero/`
New path: `/etc/airflow/data_scripts/gis/school_safety_zones`

### LINKED ISSUE #512: Flagging invalid SSZ records
Flagging invalid school safety zone records by marking the corresponding task as `failed`

### PR DESCRIPTION
**Pull request by tahaislam**
## What this pull request accomplishes:

- Updated the path to `schools.py` script that is imported from `vz_google_sheets` DAG
  - now the path is independent of the repo path 
- Marking any year with invalid records as `failed`
  - while continuing inserting all other valid records

## Issue(s) this solves:

- resolves #511 
- resolves #512

### CODE REVIEW

**Comment by radumas** [comment]
My two comments based on today's experience with the following log lines as example:

```
ERROR - WYSS : outside Toronto
ERROR - An error occurs at [row]
```

1. Can the first line include the coordinates or whatever that failed? "cannot be transformed into a date" is an example that includes it (although whatever failed should also be quoted).
2. the second line should read "on row: [row]" or something similar to link the error message with the more complete context of where the erroneous data point is. I initially thought the two lines weren't related.

File: gis/school_safety_zones/schools.py [hunk 1]
```diff
@@ -231,6 +236,8 @@ def pull_from_sheet(con, service, year, spreadsheet, *args):
             cur.execute(truncate)  
             execute_values(cur, query, rows)
     LOGGER.info('Table %s is done', table)
+    if any_error:
+        raise AirflowFailException
```
Comment (radumas) [inline gis/school_safety_zones/schools.py:240]: Is it possible to pass text to this? Such that the log line would be something like

```
[2022-11-28 13:56:12,548] {taskinstance.py:1150} ERROR - invalid rows pulled. See errors documented above.
```

**Review by radumas — CHANGES_REQUESTED** [review]

**Comment by radumas** [comment]
Reporting it back seems simpler at the moment, what do you think?
Wish there was an easy validation that could work in Sheets formulas

**Review by chmnata — CHANGES_REQUESTED** [review]
Looks good and well documented :meow_party:!! One general question that is not related to the tagged issues: We are still using a `key.json` from joven's home directory (which I thought we killed?), and a db config from cathy's home, should we change this? They are in line 258 and 264 respectively in `schools.py`. Cathy's email is also listed for alert emeows.

**Comment by tahaislam** [comment]
> Looks good and well documented :meow_party:!! One general question that is not related to the tagged issues: We are still using a `key.json` from joven's home directory (which I thought we killed?), and a db config from cathy's home, should we change this? They are in line 258 and 264 respectively in `schools.py`. Cathy's email is also listed for alert emeows.

good catch. we can make it more generic and read these variables as arguments to the script, if we still need to keep the file's `__main__` environment. These hard-coded local paths are only used when `schools.py` run on its own outside Airflow. Hence, it doesn't have access to Airflow's variables and connections (right?)

File: gis/school_safety_zones/schools.py [hunk 2]
```diff
@@ -204,10 +208,12 @@ def pull_from_sheet(con, service, year, spreadsheet, *args):
                     LOGGER.info('Reading %s columns of data from Google Sheet', len(row))
```
[the remark this answers is not in the document]
Comment (tahaislam) [reply gis/school_safety_zones/schools.py:208]: agree, printing the same statement doesn't add any value, especially that all rows almost have the same number of columns. I guess the message was intended initially to print out the number of columns in each processed row to identify any outliers. But this's not needed now, I guess

**Comment by chmnata** [comment]
> > Looks good and well documented :meow_party:!! One general question that is not related to the tagged issues: We are still using a `key.json` from joven's home directory (which I thought we killed?), and a db config from cathy's home, should we change this? They are in line 258 and 264 respectively in `schools.py`. Cathy's email is also listed for alert emeows.
> 
> good catch. we can make it more generic and read these variables as arguments to the script, if we still need to keep the file's `__main__` environment. These hard-coded local paths are only used when `schools.py` run on its own outside Airflow. Hence, it doesn't have access to Airflow's variables and connections (right?)

hmmm yea I'm not sure if we need it then, since we can run the script from /airflow and those paths doesn't exist

**Review by chmnata — CHANGES_REQUESTED** [review]

**Comment by radumas** [comment]
CodeFactor makes a good point this is a dangerous "bare" `except:` 

https://github.com/CityofToronto/bdit_data-sources/blob/c5cf17ea11b8a6c513ce3a1ad487ee2d556d2eb7/gis/school_safety_zones/schools.py#L108-L122

File: gis/school_safety_zones/schools.py [hunk 3]
```diff
@@ -153,7 +154,7 @@ def is_int(n):
 
 #--------------------------------------------------------------------------------------------------
 
-def pull_from_sheet(con, service, year, spreadsheet, *args):
+def pull_from_sheet(con, service, year, spreadsheet, **kwargs):
```
Comment (radumas) [inline gis/school_safety_zones/schools.py:157]: Is this function actually receiving keyword arguments?

Comment (tahaislam) [reply gis/school_safety_zones/schools.py:157]: yes, it's used to send some variables over XCom to the slack alert function

**Review by radumas — COMMENTED** [review]
This is awesome! We should have a lightning talk about this use of xcom.

Do you think we should set up the code with task_instance and AirflowFailException such that it could run without being in the Airflow environment or is that irrelevant going forward for this pipeline?

**Comment by tahaislam** [comment]
> This is awesome! We should have a lightning talk about this use of xcom.
> 
> Do you think we should set up the code with task_instance and AirflowFailException such that it could run without being in the Airflow environment or is that irrelevant going forward for this pipeline?

Maybe. This would require having a small intermediate function in the DAG calling `pull_from_sheet` and sending whatever needed over xcom and returning the appropriate exception when needed. Or is there another way to do it?

File: gis/school_safety_zones/schools.py (hunk 1 above)
Comment (radumas) [reply gis/school_safety_zones/schools.py]: @tahaislam do you think there should be something logged before the `AirflowFailException` would appear?

Comment (tahaislam) [reply gis/school_safety_zones/schools.py]: This's already resolved in this [commit](https://github.com/CityofToronto/bdit_data-sources/commit/0fb49c656f1cc1cba17930e138b897fd7d6319d7), but unfortunately, it's not showing here in this PR
https://github.com/CityofToronto/bdit_data-sources/blob/4eca6882859696f2794a5e5e0ab39ef029ee2618/gis/school_safety_zones/schools.py#L246

There're more commits in this branch that are not showing in this PR. And I don't know why this's happenning

**Review by radumas — APPROVED** [review]
I see the changes!

**Comment by radumas** [comment]
@chmnata can you also approve :)

File: dags/vz_google_sheets.py [hunk 4]
```diff
@@ -90,12 +102,13 @@ def task_fail_slack_alert(context):
 DEFAULT_ARGS = {
     'owner': 'cnangini',
```
Comment (chmnata) [inline dags/vz_google_sheets.py:103]: let's change this to itaha

**Review by chmnata — APPROVED** [review]

**Comment by chmnata** [comment]
Possible closing of this #434

File: dags/vz_google_sheets.py (hunk 4 above)
Comment (tahaislam) [reply dags/vz_google_sheets.py]: Done

**Pull request merged**